For the Enhancement of converters of QuadraticProgram

a-matsuo opened the pull request:
### Summary
This closes #1037

### Details and comments
Changed the following things based on #1037 

- introduced `QuadraticProgramConverter` which is an abstract class for converters
- renamed `encode`/`decode` to `convert`/`interpret`
- moved all parameters to `__init__`
- created setter/getter for converter parameters)
- enabled “auto_penalty” (#1043)
- error messages are more informative now
- added `to_ising()` and `from_ising` functions to `QuadraticProgram`, and then, removed `QuadraticProgramToIsing` class and `IsingToQuadraticProgram` class.

a-matsuo commented:
The arguments of the encode method of `LinearEqualityToPenalty` are moved to its constrictor in #1043.

woodsp-ibm commented on qiskit/optimization/converters/inequality_to_equality.py:
```diff
@@ -42,26 +42,30 @@ class InequalityToEquality:
-    def encode(
-            self, op: QuadraticProgram, name: Optional[str] = None, mode: str = 'auto'
-    ) -> QuadraticProgram:
+    def encode(self, op: QuadraticProgram) -> QuadraticProgram:
```
Is the intent to do these changes with no deprecation at all? The release will have been out for a several weeks and users may have code passing parameters on the encode.

a-matsuo replied:
Yes, It's my intent. Shall I duplicate the `encode` method and add some deprecation message to old one?

t-imamichi commented on qiskit/optimization/converters/inequality_to_equality.py:
```diff
@@ -42,26 +42,30 @@ class InequalityToEquality:
     _delimiter = '@'  # users are supposed not to use this character in variable names
 
-    def __init__(self) -> None:
+    def __init__(self, name: Optional[str] = None, mode: str = 'auto') -> None:
```
We need to change the default behavior carefully. Do we have to make 'auto' mode default?

a-matsuo replied:
Not have to, but 'auto' seems ok for the default mode and it was the default mode before this change, so I chose 'auto'.

woodsp-ibm commented on qiskit/optimization/converters/quadratic_program_to_qubo.py:
```diff
@@ -104,7 +109,7 @@ def get_compatibility_msg(problem: QuadraticProgram) -> str:
         # initialize message
         msg = ''
-
+        print(problem)
```
I guess this is a left over from debugging? We should not have print() calls in the code. A logger with the logging function though can remain if it is felt this is useful to have

woodsp-ibm commented on qiskit/optimization/converters/quadratic_program_to_qubo.py:
```diff
@@ -115,6 +120,10 @@ def get_compatibility_msg(problem: QuadraticProgram) -> str:
             msg += 'Only linear equality constraints are supported.'
```
This message does not have a trailing whitespace, that the others do, presumably so they can be concatenated and still read easily.

woodsp-ibm commented on qiskit/optimization/converters/inequality_to_equality.py:
```diff
@@ -363,3 +366,28 @@ def _decode_var(self, names, vals) -> List[int]:
         return new_vals
+
+    def is_compatible_with_integer_slack(self, prog: QuadraticProgram) -> bool:
+        """Checks whether the integer slack mdoe can be used for a given problem.
```
```suggestion
        """Checks whether the integer slack mode can be used for a given problem.
```

t-imamichi commented on qiskit/optimization/converters/inequality_to_equality.py:
```diff
@@ -347,19 +338,83 @@ def decode(self, result: OptimizationResult) -> OptimizationResult:
             new_vals.append(sol[x.name])
         return new_vals
+
+    def is_compatible_with_integer_slack(self, problem: QuadraticProgram) -> bool:
```
Is it necessary? If so, you need to write a unit test too.

t-imamichi commented on qiskit/optimization/converters/inequality_to_equality.py:
```diff
@@ -234,12 +232,11 @@ def _add_auto_slack_var_linear_constraint(self, linear, sense, rhs, name):
     def _add_integer_slack_var_quadratic_constraint(self, linear, quadratic, sense, rhs, name):
-        # If a coefficient that is not integer exist, raise an error
+        # If a coefficient that is not integer exist, raise error
         if any(
```
Similar float check appears often in the class. So, I suggest to introduce a static method.
```python
@staticmethod
    def _contains_any_float_value(values: List[Union[int, float]]) -> bool:
        return any(isinstance(val, float) and not val.is_integer() for val in values)
```

t-imamichi commented on qiskit/optimization/converters/integer_to_binary.py:
```diff
@@ -171,11 +182,16 @@ def _encode_quadratic_coefficients_dict(self, coefficients: Dict[Tuple[str, str]
+            quadratic,
+            quadratic_linear,
+            quadratic_constant,
+        ) = self._convert_quadratic_coefficients_dict(
```
You don't need to include '()'. The previous line does not include '()'.
Just, `quadratic, quadratic_linear, quadratic_constant = self._convert_quadratic_coefficients_dict(...)`

t-imamichi commented on qiskit/optimization/converters/integer_to_binary.py:
```diff
@@ -188,40 +204,50 @@ def _substitute_int_var(self):
+                quadratic,
+                quadratic_linear,
+                quadratic_constant,
+            ) = self._convert_quadratic_coefficients_dict(constraint.quadratic.to_dict())
```
Same as https://github.com/Qiskit/qiskit-aqua/pull/1061/files#r452689525

t-imamichi commented:
It should inherit `QuadraticProgramConverter`.
https://github.com/Qiskit/qiskit-aqua/blob/dbf881b4516e70f72023a910028ab878657337bf/qiskit/optimization/converters/linear_equality_to_penalty.py#L32

woodsp-ibm commented on qiskit/optimization/converters/ising_to_quadratic_program.py:
```diff
@@ -173,3 +178,42 @@ def _create_qubo_matrix(self):
+        return self._name
+
+    @name.setter  # type:ignore
+    def name(self, name: Optional[str]) -> None:
```
Should the name here be Optional? In the init it defaults to empty string ' ' and the getter here does not support return of Optional

stefan-woerner replied:
see later comment on `name`.

woodsp-ibm commented on qiskit/optimization/converters/quadratic_program_converter.py:
```diff
@@ -0,0 +1,39 @@
+    @abstractmethod
+    def convert(self, problem: QuadraticProgram) -> QuadraticProgram:
+        """
+        convert an QuadratciProgram into another form
```
```suggestion
        Convert a QuadraticProgram into another form
```

woodsp-ibm commented on qiskit/optimization/converters/quadratic_program_to_ising.py:
```diff
@@ -48,16 +47,26 @@ def encode(self, op: QuadraticProgram) -> Tuple[OperatorBase, float]:
-                                          'The method supports only model with no constraints.')
+                                          'The method supports only model with no constraints. '
+                                          'Use a QuadraticProgramToQubo converter. '
+                                          'It converters inequality constraints to equality '
```
```suggestion
                                          'It converts inequality constraints to equality '
```

woodsp-ibm commented on qiskit/optimization/converters/quadratic_program_converter.py:
```diff
@@ -0,0 +1,39 @@
+from ..problems.quadratic_program import QuadraticProgram
+
+
+class QuadraticProgramConverter(ABC):
```
As 'name' seems to be a common aspect would it make sense to have a constructor here that took and stored the name so this class would have the setter/getter for name too. I see a lot of setter/getter pairs for name hence wondering if more common function can go to this class.

stefan-woerner replied:
I think we need to discuss about `name` again. This seems to be an exception to the rule that all parameters should be moved to `__init__`. The given `name` is the name of the newly created `QuadraticProgram`. Thus, setting this as a parameter of the converter object doesn't seem to make sense (all problems would get the same name). It should be an optional argument to `convert` that is `None` by default, which means the new problem has the same name as the old problem. This also implies that we don't need setters/getters for `name`.

t-imamichi replied:
Because algorithms can modify the name of `QuadraticProgram`, it might be good not to touch name in converters to be simple.

stefan-woerner replied:
@t-imamichi good for me.

woodsp-ibm commented:
Can you also please do a reno yaml file to describe the changes that have been made for the convertors (see CONTRIBUTING.md file for further info on reno if you need it)

a-matsuo commented:
@woodsp-ibm @t-imamichi @stefan-woerner 
Thank you very much for your advices. I fixed spells, and moved `name` argument of converters from constructors to `convert` method. I also added a release note by reno.

woodsp-ibm commented on qiskit/optimization/converters/inequality_to_equality.py:
```diff
@@ -71,13 +72,18 @@ def encode(
         self._dst = QuadraticProgram()
+
+        # set a problem name
         if name:
```
An empty string, if the user provided it will evaluate False, as does the default value of None. Should this test be `if name is not None` to allow empty string to be set as a problem name. (I see the same test in other convertors of course)

a-matsuo commented:
I missed comments about `name` arguments from @stefan-woerner @t-imamichi. 
I'm going to remove `name` arguments from converters right away.

woodsp-ibm commented on qiskit/optimization/converters/quadratic_program_converter.py:
```diff
@@ -0,0 +1,40 @@
+    @abstractmethod
+    def convert(self, problem: QuadraticProgram, name: Optional[str]) -> QuadraticProgram:
+        """
+        Convert an QuadratciProgram into another form
```
```suggestion
        Convert a QuadraticProgram into another form
```
Spelling

a-matsuo commented:
Hi @woodsp-ibm 
I finished implementation. Could you please check the code?
By the way, thank you very much for always checking codes. I appreciate a lot!

woodsp-ibm commented on qiskit/optimization/converters/quadratic_program_converter.py:
```diff
@@ -29,7 +28,7 @@ class QuadraticProgramConverter(ABC):
     def convert(self, problem: QuadraticProgram) -> QuadraticProgram:
         """
-        Convert an QuadratciProgram into another form
+        Convert a QuadratciProgram into another form
```
```suggestion
        Convert a QuadraticProgram into another form
```
This `QuadratciProgram` still seems to be spelled wrong

woodsp-ibm commented on qiskit/optimization/converters/ising_to_quadratic_program.py:
```diff
@@ -68,8 +69,10 @@ def encode(self, qubit_op: Union[OperatorBase, WeightedPauliOperator], offset: f
+            )
 
         self._qubit_op = qubit_op
         self._offset = copy.deepcopy(offset)
```
This deepcopy for offset, which is a float, - was that really intended?

woodsp-ibm commented on qiskit/optimization/converters/ising_to_quadratic_program.py:
```diff
@@ -173,3 +176,24 @@ def _create_qubo_matrix(self):
+        """
+        return self._linear
+
+    @linear.setter  # type:ignore
```
Is this pylint `type: ignore` on the setter needed? I see it elsewhere on another setter so am wondering if it was inadvertently copied from somewhere where it was needed.

stefan-woerner commented:
This looks good to me. I have two last points on refactoring that I'd like to discuss:

1. Should we add `to_ising` and `from_ising` directly in the `QuadraticProgram`, as we do with the `to/from_docplex`? This would remove the Ising-converters from the converter directory, which would be nice as they do not implement the base class.
2. Should we create a folder `grover_optimizer` for the `GroverOptimizer` and move it together with the `NegValueOracle` there (like e.g. for QAOA under aqua.algorithms)? Same rationale, it does not implement the base class, and, in the long run, we want to move this one as an arithmetic circuit to the circuit library instead of having it as a converter.

Once we agreed on how to proceed with 1. and 2. I am good to go.

t-imamichi commented:
Embedding `to_ising` and `from_ising` in `QuadraticProgram` makes sense. Another option is to make `IsingHamiltonian` class as follows. Pros: keeping the simplicity of QP, Cons: introducing another class `IsingHamiltonian`.
```python
class IsingHamiltonian:
  def __init__(op: Optional[OperatorBase] = None):
     self._qubit_op =  op
     ...

  @property
  def qubit_op(self) -> OperatorBase:
    return self._qubit_op
  
  def to_qp(self) -> QuadraticProgram:
    ...

  # from_qp is a class method or static method
  def from_qp(self, qp: QuadraticProgram):
    ...
  
  @staticmethod
  def from_qp(qp: QuadraticProgram) -> IsingHamiltonian:
    ...
```

woodsp-ibm commented:
Is there a distinction here between having the to_ising in QP but not say a to_oracle method. Or maybe the latter is only a partial representation. I guess there are two types of entities in convertors - those that transforms QP from one form of a QP to another, along with equivalent for handling result. And those which translate it into a form more suitable for processing by some algorithm - these being the ising and oracle forms. 
For GroverOptimizer - I am not sure of the final split between that module and the NegValueOracle as the latter convertor acts on a QP whereas as circuit going to the library I would though would have more fundamental input. If the convertor is felt to be more of a general purpose circuit, maybe it could go to Aqua circuits for now since we have been migrating form these to the library. Otherwise would we end up with GroverOptimizer just in a folder of its own and want to move it yet again in the future?

stefan-woerner commented:
There are some differences between `to_ising()` and `to_oracle()`:
- There is a standard and straight forward way to translate between Ising Hamiltonian and Quadratic Program, while there are plenty of possibilities to implement the oracle, we just have only one implemented. 
- The oracle converter is reusable in a broader context as an arithmetic circuit. But you are right @woodsp-ibm, it would not take a `QuadraticProgram` as input, but probably a cost matrix defining a quadratic objective function on binary variables (x^T*Q*x).
Thus, my suggestion above to move the oracle-converter out of `converters` and somewhere else to prepare moving it to the circuit library, and to include the `to/from_ising()` in the `QuadraticProgram`. I think that's the cleanest solution. 
Where exactly we move  the oracle-converter is a good question. @woodsp-ibm maybe the `aqua.circuits` would be a good intermediate solution. I need to check with @Cryoris, maybe we can also directly move it to `qiskit.circuit.library`.

woodsp-ibm commented:
So for the oracle based solutions it sounds perhaps as if the QuadraticProgram could have a to_cost_matrix() that would be a common aspect - this then being used for potentially different Oracles implementations that take such as input however they are built out. So given a Grover with one of these Oracles the optimization module can plug in the cost matrix in an analogous manner to the operator for the min eigen solvers.

a-matsuo commented:
Adding `to/from_ising()` in a `QuadraticProgram` makes sense to me as well. AIright, I will add `to_ising` and `from_ising` in the `QuadraticProgram`.

stefan-woerner commented:
Thanks @a-matsuo!

@woodsp-ibm The corresponding circuit would take a quadratic, linear, and constant part. These are all already provided by `quadratic_program.objective.constant/linear/quadratic (.to_array())`.

a-matsuo commented:
I added `to_ising()` and `from_ising()` functions to `QuadraticProgram` class, and then, removed `QuadraticProgramToIsing` class and `IsingToQuadraticProgram` class.
@stefan-woerner 
What shall I do with a file for `QuadraticProgramToNegativeValueOracle`? Shall I move to `aqua.circuits` or can we directly move to `qiskit.circuit.library`?

stefan-woerner commented:
@a-matsuo thanks! @Cryoris will take care of the `QuadraticProgramToNegativeValueOracle` and move it to `qiskit.circuit.library`.

a-matsuo commented:
Sorry for taking time. It seems that finally all the tests is passed, and it's ready to be merged.

woodsp-ibm commented:
The `QuadraticProgramToNegativeValueOracle` change has yet to be done right - you say this PR is ready, that part of the refactor will be part of a different PR now?

a-matsuo commented:
I thought we are not moving `QuadraticProgramToNegativeValueOracle` for now, because some refactors are needed before moving it.
 
@stefan-woerner 
Is my understanding correct or are you planning to move `QuadraticProgramToNegativeValueOracle` right now?

stefan-woerner commented:
@a-matsuo @woodsp-ibm We should move this converter too. @Cryoris is looking into this to integrate it in the circuit library and move the remaining logic to the `GroverOptimizer`.

a-matsuo commented:
Ah, I see. Let's wait for that.

Cryoris commented on qiskit/optimization/converters/integer_to_binary.py:
```diff
@@ -54,15 +55,14 @@ def __init__(self) -> None:
+
         # e.g., self._conv = {x: [('x@1', 1), ('x@2', 2)]}
 
-    def encode(self, op: QuadraticProgram, name: Optional[str] = None) -> QuadraticProgram:
```
It might be good to deprecate these methods instead of just removing them since the optimization module is already released since 3 months and might be used by several people. Could you add deprecation warnings saying that the new method is now `convert`? See e.g. here for how deprecation warnings usually look like: https://github.com/Qiskit/qiskit-aqua/blob/c343e52597e11a08bb20eadfff11bf93fab974e8/qiskit/aqua/components/variational_forms/ry.py#L94

You could also move that directly inside the `QuadraticProgramConverter` base class (same for `decode`/`interpret`) so that you don't have to write it for every converter.

Cryoris commented on qiskit/optimization/converters/integer_to_binary.py:
```diff
@@ -188,40 +188,48 @@ def _substitute_int_var(self):
-    def decode(self, result: OptimizationResult) -> OptimizationResult:
-        """Convert the encoded problem (binary variables) back to the original (integer variables).
+    def interpret(self, result: OptimizationResult) -> OptimizationResult:
+        """
+        Convert back the converted problem (binary variables) to the original (integer variables).
```
The documentation guidelines are w/o linebreak here

```suggestion
        """Convert back the converted problem (binary variables) to the original (integer variables).
```

Cryoris commented on qiskit/optimization/converters/linear_equality_to_penalty.py:
```diff
@@ -25,31 +25,29 @@
     """Convert a problem with only equality constraints to unconstrained with penalty terms."""
 
-    def __init__(self, penalty: Optional[float] = None, name: Optional[str] = None):
+    def __init__(self, penalty: Optional[float] = None):
```
```suggestion
    def __init__(self, penalty: Optional[float] = None) -> None:
```

Cryoris commented on qiskit/optimization/converters/linear_equality_to_penalty.py:
```diff
@@ -25,31 +25,29 @@
         Args:
-            op: The problem to be solved, that does not contain inequality constraints.
-
+            problem: The problem to be solved, that does not contain inequality constraints.
         Returns:
```
Missing linebreak 
```suggestion
            problem: The problem to be solved, that does not contain inequality constraints.
            
        Returns:
```

Cryoris commented on qiskit/optimization/converters/linear_equality_to_penalty.py:
```diff
@@ -58,22 +56,16 @@ def encode(self, op: QuadraticProgram) -> QuadraticProgram:
         # create empty QuadraticProgram model
-        self._src = copy.deepcopy(op)  # deep copy
-        self._dst = QuadraticProgram()
+        self._src = copy.deepcopy(problem)  # deep copy
```
This obviously does deepcopy 🙂  (you didn't add it but we could still remove this comment)
```suggestion
        self._src = copy.deepcopy(problem) 
```

Cryoris commented on qiskit/optimization/problems/quadratic_program.py:
```diff
@@ -851,6 +853,216 @@ def substitute_variables(
+        """Return the Ising Hamiltonian of this problem.
+
+        Returns:
+            The qubit operator of the problem and the shift value.
```
Could you add an additional sentence explaining what the shift value is? This might not be clear to everyone.

Cryoris commented on qiskit/optimization/problems/quadratic_program.py:
```diff
@@ -851,6 +853,216 @@ def substitute_variables(
+        # initialize Hamiltonian.
+        num_nodes = self.get_num_vars()
+        pauli_list = []
+        shift = 0
```
In `from_ising` this variable is called `offset`, could you give them both the same name? (Either shift or offset)

Cryoris commented on qiskit/optimization/problems/quadratic_program.py:
```diff
@@ -851,6 +853,216 @@ def substitute_variables(
+                'operator in the ListOp separately.'
+            )
+
+        # Create `QuadraticProgram`
```
```suggestion
        # add binary variables
```

a-matsuo commented:
Hi @Cryoris,
Thank you for your comments. I fixed all of them. I also added deprecated messages for `encode/decode` as @woodsp-ibm mentioned as well since it took longer than originally planned.

Cryoris commented on qiskit/optimization/converters/quadratic_program_converter.py:
```diff
@@ -0,0 +1,63 @@
+
+        raise NotImplementedError
+
+    @abstractmethod
```
If this is abstract we still have to implement it for every class, we can just implement it here to call `convert`.
```suggestion
```

Cryoris commented on qiskit/optimization/converters/quadratic_program_converter.py:
```diff
@@ -0,0 +1,63 @@
+                      'than 3 months after the release. You should use '
+                      'qiskit.optimization.converters.QuadraticProgramConverter.convert() '
+                      'instead.',
+                      DeprecationWarning, stacklevel=2)
+
```
I think you can call `convert` directly here it will then call the `convert` function of the correct subclasses.
```suggestion
                      DeprecationWarning, stacklevel=2)
        return self.convert(problem)

```

Cryoris commented on qiskit/optimization/converters/quadratic_program_converter.py:
```diff
@@ -0,0 +1,63 @@
+                      'instead.',
+                      DeprecationWarning, stacklevel=2)
+
+    @abstractmethod
```
```suggestion
```

Cryoris commented on qiskit/optimization/converters/quadratic_program_converter.py:
```diff
@@ -0,0 +1,63 @@
+                      'than 3 months after the release. You should use '
+                      'qiskit.optimization.converters.QuadraticProgramConverter.interpret() '
+                      'instead.',
+                      DeprecationWarning, stacklevel=2)
```
```suggestion
                      DeprecationWarning, stacklevel=2)
        return self.interpret(result)
```

woodsp-ibm replied:
I guess its ok to have these deprecation wrappers in the base class. Of the entire refactor this covers some part of it but I guess its better to have some. Though we had discussed and I think since it could not all be covered the thinking had not been to try. The encode works as long as no-one passed any of the defaulted parameters into it in the past since these got moved off to constructor. But I guess it will at least give some indications and may well work for most cases. For sure the encode/decode methods need to go as they are not covered by any unit tests.

stefan-woerner commented on qiskit/optimization/algorithms/grover_optimizer.py:
```diff
@@ -228,10 +228,10 @@ def solve(self, problem: QuadraticProgram) -> OptimizationResult:
         result = OptimizationResult(x=opt_x, variables=problem.variables, fval=fval,
                                     results={"grover_results": grover_results,
-                                             "qubo_converter": qubo_converter})
+                                             "qubo_converter": self._qubo_converter})
```
insert a copy of the converter?

stefan-woerner replied:
(for all optimizers)
